src/lib/gpm.c

```diff
@@ -0,0 +1,52 @@
+#include "internal.h"
+#ifdef USE_GPM
+#undef buttons // defined by both term.h and gpm.h, ugh
+#include <lib/gpm.h>
+#include <gpm.h>
+
+static Gpm_Connect gpmconn;    // gpm server handle
+
+int gpm_connect(tinfo* ti){
+  (void)ti;
+  gpm_zerobased = 1;
+  gpmconn.eventMask = ~0;
+  gpmconn.defaultMask = 0;
+  gpmconn.minMod = 0;
+  gpmconn.maxMod = ~0;
+  if(Gpm_Open(&gpmconn, 0) == -1){
+    logerror("couldn't connect to gpm");
+    return -1;
+  }
+  loginfo("connected to gpm on %d\n", gpm_fd);
+  return gpm_fd;
+}
+
+int gpm_read(tinfo* ti, ncinput* ni){
+  (void)ti;
+  (void)ni;
+  return -1;
+}
+
+int gpm_close(tinfo* ti){
+  (void)ti;
+  Gpm_Close();
+  memset(&gpmconn, 0, sizeof(gpmconn));
+  return 0;
+}
+#else
+int gpm_connect(tinfo* ti){
+  (void)ti;
+  return -1;
+}
+
+int gpm_read(tinfo* ti, ncinput* ni){
```

Comment:
Some GPM code examples:

Setup and teardown: https://gitlab.com/klamonte/lcxterm/-/blob/master/source/screen.c#L1491

Knowing when to read: https://gitlab.com/klamonte/lcxterm/-/blob/master/source/main.c#L1290

The actual reading code, whose sole purpose is to re-encode GPM events into X10 mouse protocol:

https://gitlab.com/klamonte/lcxterm/-/blob/master/source/input.c#L1635

Reply:
interesting. so lxcterm is basically a linux pure vga/fbterm/kms console ... augmenter? (and also for terminal emulators, i see, but interested in the former aspect atm)

so normally i've got getty -> login -> shell, and the shell is talking directly to the tty. in this case, lxcterm is talking to the tty, and passing things on to the linux console? i don't see how that can achieve effects beyond what the linux console offers you already? like...

> Since LCXterm uses ncurses for its display, it may lack features (such as 24-bit RGB color) that you need.

sure, but wouldn't you be missing true 24-bit RGB anyway by dint of using glyphs on the linux console, which to the best of my knowledge supports either 8 colors with 512 glyphs in the font, or 8+8 via bold with 256 glyphs in the font. it interprets 24-bit sequences but high filters that into the base 8 or 16:

>        38        256/24-bit  foreground  color follows, shoehorned into 16 basic colors (before Linux 3.16: set underscore  on,  set default foreground color)

so whether ncurses is between you and the linux console or not, you're not getting more than 8/16 colors without drawing glyphs yourself, right? likewise you're stuck with the console's dystopian garbage font (btw, in one of my more aspy bits of code, we [reprogram the console font](https://github.com/dankamongmen/notcurses/blob/master/src/lib/linux.c)).

i'm surprised ncurses really saved you much effort on this project.

Reply:
ideally i'll be pumping this up to UTF-8 mouse protocol, which is what i request from all terminals. that shouldn't introduce any conceptual complexity, though.

Reply:
btw assuiming the PM/ASC/DCS patch goes through without any significant pushback and annoyance, one of my next plans is to implement true 8bpc glyphs when available, i.e. directly in the kernel as opposed to needing `fbterm` or `kmscon` jankage.

Reply:
> interesting. so lxcterm is basically a linux pure vga/fbterm/kms console ... augmenter? (and also for terminal emulators, i see, but interested in the former aspect atm)
> 
> so normally i've got getty -> login -> shell, and the shell is talking directly to the tty. in this case, lxcterm is talking to the tty, and passing things on to the linux console? i don't see how that can achieve effects beyond what the linux console offers you already? like...

LCXterm is a stripped-down Qodem, keeping just a few things available for any terminal.  Take Zmodem: there have been feature requests for Zmodem support for putty, alacritty, and xterm.js at the least.  (Konsole actually supports it by way of piping to lrzsz.)  But most terminals won't do it, because Zmodem royally sucks as a protocol and very few people use it today, despite the handiness of having any already-running ssh tty being able to double as sftp/scp.  LCXterm can get you Zmodem, even if your primary terminal will never do it.

When I worked on oceanographic buoys, it was a PC-104 form factor with serial port or rlogin/ssh console. The laptop I used was raw Linux console with 12 ttys (gettys on every function key).  I used Qodem on several of the ttys to give me: a long persistent scrollback, Zmodem file transfer, and keyboard macros (so I could just press say F8 and get "power_sensors.pl -u " filled in, and I just add "S2" to mean put power on to sensor number 2).  Later on I was doing devops, and used Qodem under xterm with keyboard macros for tmux (vertical/horizontal split, next pane, attach, detach), and it made life quite a bit easier and faster for me.

Qodem is where I learned the basics of ncurses, hence its continued use for LCXterm.  But now I know enough to skip the ncurses part for the future. :)  But also since my memory sucks, if I am stuck doing ncurses again someday: https://gitlab.com/klamonte/terminal-snippets/raw/master/README.md

But maybe next time it's notcurses instead. :-)